projects/batfish/src/main/java/org/batfish/bdp/AbstractRib.java

```diff
@@ -78,6 +85,68 @@ public void collectRoutes(Set<R> routes) {
       routes.addAll(_routes);
     }
 
+    public boolean containsRoute(
+        R route, BitSet bits, int prefixLength, int firstUnmatchedBitIndex) {
+      if (prefixLength == _prefix.getPrefixLength()) {
+        return _routes.contains(route);
+      } else {
+        boolean currentBit = bits.get(firstUnmatchedBitIndex);
+        if (currentBit) {
+          if (_right == null) {
+            return false;
+          } else {
+            Prefix rightPrefix = _right._prefix;
+            int rightPrefixLength = rightPrefix.getPrefixLength();
+            Ip rightAddress = rightPrefix.getAddress();
+            BitSet rightAddressBits = rightAddress.getAddressBits();
+            int nextUnmatchedBit;
+            boolean currentAddressBit = false;
+            boolean currentRightAddressBit;
+            for (nextUnmatchedBit = firstUnmatchedBitIndex + 1;
+                nextUnmatchedBit < rightPrefixLength && nextUnmatchedBit < prefixLength;
```

Comment:
I'm not quite understanding how the case is handled when `prefixLength` > `rightPrefixLength`. It looks like it will always return false, because even if everything matches we'll exit the for loop, fail the check at L114, and then return false.

If this is desired, just have that return earlier?

I am likely missing something.

Reply:
This code was trivially modified from the `mergeRoute` code which modifies the underlying DDNF-like structure containing the routes. This code itself was itself somewhat brutishly modified from the original trie-maintaining code before the structure was changed. There is likely quite a bit of room for improvement via careful analysis and tests in both `mergeRoute` and `containsRoute`.
I will say that at the time I wanted to avoid further function calls (helpers) because all this code is in a critical section (though that might have been overly paranoid). 

Anyway the  `return false` on L117 would in `mergeRoute` instead be code that performs modification. That modification depends on how far we get along `rightAddressBits`, regardless of whether `prefixLength` is larger than `rightPrefixLength`. However in this case since we don't care how far we get when `prefixLength` is larger than `rightPrefixLength`, the code can be simplified.

Due to the large amount of optimization we can do on this class, I'd personally prefer to just file a ticket and defer to a later PR.